lib/web_ui/dev/test_runner.dart

```diff
@@ -0,0 +1,178 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io' as io;
+
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+import 'package:test_core/src/runner/hack_register_platform.dart' as hack; // ignore: implementation_imports
+import 'package:test_api/src/backend/runtime.dart'; // ignore: implementation_imports
+import 'package:test_core/src/executable.dart' as test; // ignore: implementation_imports
+
+import 'test_platform.dart';
+import 'environment.dart';
+
+Future<void> runTests() async {
+  await _buildHostPage();
+  await _buildTests();
+
+  if (environment.targets.isEmpty) {
+    await _runAllTests();
+  } else {
+    await _runSingleTest(environment.targets);
+  }
+}
+
+Future<void> _runSingleTest(List<String> targets) async {
+  await _runTestBatch(targets, concurrency: 1, expectFailure: false);
+  _checkExitCode();
+}
+
+Future<void> _runAllTests() async {
+  final io.Directory testDir = io.Directory(path.join(
+    environment.webUiRootDir.path,
+    'test',
+  ));
+
+  // Separate screenshot tests from unit-tests. Screenshot tests must run
+  // one at a time. Otherwise, they will end up screenshotting each other.
+  // This is not an issue for unit-tests.
+  const String failureSmokeTestPath = 'test/golden_tests/golden_failure_smoke_test.dart';
+  final List<String> screenshotTestFiles = <String>[];
+  final List<String> unitTestFiles = <String>[];
+
+  for (io.File testFile in testDir.listSync(recursive: true).whereType<io.File>()) {
+    final String testFilePath = path.relative(testFile.path, from: environment.webUiRootDir.path);
+    if (!testFilePath.endsWith('_test.dart')) {
+      // Not a test file at all. Skip.
+      continue;
+    }
+    if (testFilePath.endsWith(failureSmokeTestPath)) {
+      // A smoke test that fails on purpose. Skip.
+      continue;
+    }
+    if (path.split(testFilePath).contains('golden_tests')) {
+      screenshotTestFiles.add(testFilePath);
+    } else {
+      unitTestFiles.add(testFilePath);
+    }
+  }
+
+  // This test returns a non-zero exit code on purpose. Run it separately.
+  await _runTestBatch(
+    <String>[failureSmokeTestPath],
+    concurrency: 1,
+    expectFailure: true,
+  );
+  _checkExitCode();
+
+  // Run all unit-tests as a single batch.
+  await _runTestBatch(unitTestFiles, concurrency: 10, expectFailure: false);
+  _checkExitCode();
+
+  // Run screenshot tests one at a time.
+  for (String testFilePath in screenshotTestFiles) {
+    await _runTestBatch(<String>[testFilePath], concurrency: 1, expectFailure: false);
+    _checkExitCode();
```

Comment:
Can we use the return value of `_runTestBatch` instead of relying on the indirect communication between the two functions through `io.exitCode`?

Reply:
Unfortunately no. This relies on the `package:test`'s test runner API, which communicates test success and failures by setting the exit code. It kind of makes sense, since on CI you communicate via process exit codes as well. Otherwise, if some place in the code accidentally catches an exception or ignores the returned exit code the CI will incorrectly pass.